crates/services/p2p/src/discovery.rs

```diff
@@ -301,14 +273,20 @@ mod tests {
         };
 
         let listen_addr: Multiaddr = Protocol::Memory(rand::random::<u64>()).into();
-        let swarm_builder = SwarmBuilder::without_executor(
-            transport,
-            behaviour,
-            keypair.public().to_peer_id(),
-        )
-        .dial_concurrency_factor(NonZeroU8::new(1).expect("1 > 0"));
+        // let swarm_builder = SwarmBuilder::without_executor(
+        //     transport,
+        //     behaviour,
+        //     keypair.public().to_peer_id(),
+        // )
+        // .dial_concurrency_factor(NonZeroU8::new(1).expect("1 > 0"));
+
+        // let mut swarm = swarm_builder.build();
+
+        let config = libp2p_swarm::Config::with_tokio_executor()
```

Comment:
This is a guess because there is no `without_executor` concept anymore.

Reply:
For testing, I'd recommend using `libp2p_swarm_test` and the `Swarm::new_ephemeral` function.

Reply:
Okay. This `discovery` test was broken so I switched over to `libp2p_swarm_test` and that seemed to make it run better. It doesn't fail immediately, but it's still broken.

The test creates 25 nodes and tries to connect them all together starting with just one shared peer. It never seems to connect to anyone but the original peer--I can reduce the node count to 2 and the test passes, but at 3 nodes it fails.

It seems to be failing because of protocol negotiation failure:
```
2023-12-02T01:02:29.023409Z DEBUG new_established_connection{remote_addr=/memory/15897544716629232405/p2p/12D3KooWNUZev6j3zKSe8iEKuGW8jHxhGKe9QdeLxcyyWdaqgZBv id=3 peer=12D3KooWNUZev6j3zKSe8iEKuGW8jHxhGKe9QdeLxcyyWdaqgZBv}:Connection::poll: multistream_select::dialer_select: Dialer: Received rejection of protocol: /fuel/kad/test_network/kad/1.0.0    
```
and
```
2023-12-02T01:02:29.023766Z DEBUG Swarm::poll: libp2p_kad::behaviour: Request to peer in query failed with Upgrade(NegotiationFailed) peer=12D3KooWNUZev6j3zKSe8iEKuGW8jHxhGKe9QdeLxcyyWdaqgZBv query=QueryId(4)
```
I've been able to track the problem back as far as a `StreamUpgradeError` in the Connection's `negotiating_out`. I can keep looking up that way but I'm also wondering if this has something to do with the `mdns` stuff that got removed. I'm not sure--this is still a little magic to me. But given this is a pretty core function of kad, I thought you might have some ideas.

Reply:
It looks like part of the issue is that Kademlia now has "client" and "server" modes and it defaults to "client". If we change to "server", it passes for 3.

25 is still failing. It seems to break around 10 nodes (10 is flaky). Still investigating that.

Reply:
> It looks like part of the issue is that Kademlia now has "client" and "server" modes and it defaults to "client". If we change to "server", it passes for 3.

Yes. Kademlia will automatically switch to server-mode if the node learns that it has an external address. You can do that as part of listening when you set up the test-swarm:

```rust
let mut swarm1 = Swarm::new_ephemeral(|identity| { ... });
swarm1.listen().with_memory_addr_external().await;
```

Reply:
Any node that should be part of the routing table needs to be in server-mode.